### PR TITLE
Pract.04: Add 02 | Tsvetelina Mancheva

### PR DESCRIPTION
**Pull request by manchevatsveti**
#Tsvetelina Mancheva
SI 3 0MI0600239

### CODE REVIEW

File: Sem. 04/Pract. 04/Task02.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		
+		unsigned nameLen = 0;
+		nameLen = strlen(name);
+
+		this->name = new char[nameLen + 1];
+		strcpy(this->name, name);
+	}
+	void setNameFromFile(std::ifstream& ifs, unsigned nameLen) {
+		this->name = new char[nameLen+1];
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task02.cpp]: memory leak. What if we call setNameFromFile multiple times?

File: Sem. 04/Pract. 04/Task02.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		unsigned nameLen = 0;
+		nameLen = strlen(description);
+
+		this->description = new char[nameLen + 1];
+		strcpy(this->description, description);
+		this->description[nameLen] = '\0';
+	}
+
+	void setDescriptionFromFile(std::ifstream& ifs, unsigned descriptionLen) {
+		this->description = new char[descriptionLen + 1];
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task02.cpp]: memory leak

File: Sem. 04/Pract. 04/Task02.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		ifs.read(this->name, descriptionLen);
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task02.cpp]: This causes the heap corruption. You are populating the wrong array. this->name has length less than this->description and when you read the description in the name array it overflows hence heap corruption

File: Sem. 04/Pract. 04/Task02.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+		unsigned nameLen = 0;
+		nameLen = strlen(description);
+
+		this->description = new char[nameLen + 1];
+		strcpy(this->description, description);
+		this->description[nameLen] = '\0';
+	}
+
+	void setDescriptionFromFile(std::ifstream& ifs, unsigned descriptionLen) {
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task02.cpp]: this is basically the same function as setNameFromFile().Avoid code duplication

File: Sem. 04/Pract. 04/Task02.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+		this->description = new char[descriptionLen + 1];
+		ifs.read(this->name, descriptionLen);
+		this->description[descriptionLen] = '\0';
+	}
+
+	void setPoints(unsigned points) {
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task02.cpp]: setter methods are a good place to validate input

File: Sem. 04/Pract. 04/Task02.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	}
+};
+
+class Exam {
+private:
+	Task tasks[ARRAY_SIZE];
+	unsigned minPoints;
+
+public:
+	Exam() {
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task02.cpp]: this can be rewritten as `Exam() = default;`

**Review by YavorAlexandrov — CHANGES_REQUESTED** [review]

File: Sem. 04/Pract. 04/Task.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+	void setDescription(const char* description);
+	void setNameFromFile(std::ifstream& ifs, unsigned nameLen);
+	void setDescriptionFromFile(std::ifstream& ifs, unsigned descriptionLen);
+	void setPoints(unsigned points);
+
+	const char* getName() const;
+	const char* getDescription() const;
+	unsigned getPoints() const;
+
+	void print();
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task.h:24]: function should be const

File: Sem. 04/Pract. 04/Task.cpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
+#include "Task.h"
+#include <fstream>
+
+
+#pragma warning( disable : 4996)
+
+Task::Task() {
+	name = nullptr;
+	description = nullptr;
+	points = 0;
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task.cpp:10]: default values are already set in the class definition

File: Sem. 04/Pract. 04/Task.cpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+	if (this->description != nullptr) {
+		delete[] this->description;
+	}
+
+	unsigned nameLen = 0;
+	nameLen = strlen(description);
+
+	this->description = new char[nameLen + 1];
+	strcpy(this->description, description);
+	this->description[nameLen] = '\0';
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task.cpp:41]: strcpy already sets this

File: Sem. 04/Pract. 04/Task.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+unsigned Task::getPoints() const {
+	return points;
+}
+
+void Task::print() {
+	const char* name = getName();
+	unsigned nameLen = strlen(name);
+	for (int i = 0; i < nameLen; i++) {
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task.cpp:84]: simply print it as `std::cout<<name;`

File: Sem. 04/Pract. 04/Task.cpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+		std::cout << name[i];
+	}
+	std::cout << " ";
+
+	const char* description = getDescription();
+	unsigned desLen = strlen(description);
+	for (int i = 0; i < desLen; i++) {
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Task.cpp:91]: `std::cout<<description;`

File: Sem. 04/Pract. 04/Exam.h [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+	unsigned minPoints;
+
+public:
+	Exam();
+	Exam(const Task* tasks, unsigned minPoints);
+
+	void setPoints(unsigned minPoints);
+
+	void changeMin(const unsigned minPoints);
+	unsigned getMax();
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Exam.h:18]: function should be const

File: Sem. 04/Pract. 04/Exam.cpp [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+	}
+	this->minPoints = minPoints;
+}
+
+void Exam::setPoints(unsigned minPoints) {
+	this->minPoints = minPoints;
+}
+
+void Exam::changeMin(const unsigned minPoints) {
+	this->minPoints = minPoints;
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Exam.cpp:21]: validate minPoints

File: Sem. 04/Pract. 04/Exam.cpp [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+Exam::Exam(const Task* tasks, unsigned minPoints) {
+	for (int i = 0; i < ARRAY_SIZE; i++) {
+		this->tasks[i].setName(tasks[i].getName());
+		this->tasks[i].setDescription(tasks[i].getDescription());
+		this->tasks[i].setPoints(tasks[i].getPoints());
+	}
+	this->minPoints = minPoints;
+}
+
+void Exam::setPoints(unsigned minPoints) {
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Exam.cpp:16]: this is the same function as changeMin()

File: Sem. 04/Pract. 04/Exam.h [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+	void writeToFile(const char* filename) const;
+	void readFromFile(const char* filename);
+
+	void print();
```
Comment (YavorAlexandrov) [inline Sem. 04/Pract. 04/Exam.h:23]: function should be const

**Review by YavorAlexandrov — CHANGES_REQUESTED** [review]

**Comment by GeorgiTerziev02** [comment]
Due to inactivity

**Pull request closed without merging**